### PR TITLE
[FLINK-11889] Remove "stop" signal and related interfaces.

### PR DESCRIPTION
**Pull request by kl0u**
## What is the purpose of the change

This change aims at removing the "old" STOP command, as decided in the related discussion in the mailing list (please see related JIRA for a link to the discussion). In addition, after the merging of the commits related to FLIP-34 and https://issues.apache.org/jira/browse/FLINK-11458, the semantics of the STOP command have now change.

## Brief change log

- *The old "stop" REST API call returns a meaningful error code  *
- *The `StoppableStreamSource`, `StoppableStreamSourceTask`, `StoppableInvokable` and `StoppableFunction` have been removed *
- *All the code paths that make use of the above (RPCs on TaskManager and JobManager, etc) have been removed*

## Verifying this change

This change removes functionality and is "verified" by the fact that no test fails.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no although it changes a REST API call**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

R @zentol

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobCancellationHandler.java [hunk 1]
```diff
@@ -83,8 +84,7 @@ public JobTerminationHandler(
 				terminationFuture = gateway.cancelJob(jobId, timeout);
 				break;
 			case STOP:
-				terminationFuture = gateway.stopJob(jobId, timeout);
-				break;
+				throw new RestHandlerException("The \"stop\" command is deprecated. Please use \"stop-with-savepoint\" instead.", HttpResponseStatus.PERMANENT_REDIRECT);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobCancellationHandler.java]: Instead of deprecated let's say that it was removed instead.

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobCancellationHandler.java [hunk 2]
```diff
@@ -45,26 +44,28 @@
...
 			GatewayRetriever<? extends RestfulGateway> leaderRetriever,
 			Time timeout,
 			Map<String, String> headers,
-			MessageHeaders<EmptyRequestBody, EmptyResponseBody, JobTerminationMessageParameters> messageHeaders,
+			MessageHeaders<EmptyRequestBody, EmptyResponseBody, JobCancellationMessageParameters> messageHeaders,
 			TerminationModeQueryParameter.TerminationMode defaultTerminationMode) {
 		super(leaderRetriever, timeout, headers, messageHeaders);
 
-		this.defaultTerminationMode = Preconditions.checkNotNull(defaultTerminationMode);
+		this.defaultTerminationMode = checkNotNull(defaultTerminationMode);
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/JobCancellationHandler.java]: unrelated

File: flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/TerminationModeQueryParameter.java [hunk 3]
```diff
@@ -18,11 +18,11 @@
 
 package org.apache.flink.runtime.rest.messages;
 
-import org.apache.flink.runtime.rest.handler.job.JobTerminationHandler;
 import org.apache.flink.util.StringUtils;
 
 /**
- * Termination mode for the {@link JobTerminationHandler}.
+ * Termination mode.
+ * @deprecated Please use the "stop-with-savepoint" command instead.
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/rest/messages/TerminationModeQueryParameter.java]: "Only kept to detect legacy usages of the cancel/stop command."

File: flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java [hunk 4]
```diff
@@ -216,7 +215,7 @@ private RestClient createRestClient() throws ConfigurationException {
 	@Test
 	public void testJobSubmitCancelStop() throws Exception {
```
Comment (zentol) [inline flink-clients/src/test/java/org/apache/flink/client/program/rest/RestClusterClientTest.java]: rename to `testJobSubmitCancel`

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/WebFrontendITCase.java [hunk 5]
```diff
@@ -280,8 +279,8 @@ public void testStop() throws Exception {
 		final Deadline deadline = testTimeout.fromNow();
 
 		try (HttpTestClient client = new HttpTestClient("localhost", getRestPort())) {
-			// stop the job
-			client.sendPatchRequest("/jobs/" + jid + "/?mode=stop", deadline.timeLeft());
+			// cancel the job
+			client.sendPatchRequest("/jobs/" + jid + "/?mode=cancel", deadline.timeLeft());
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/WebFrontendITCase.java]: remove mode parameter

File: flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java [hunk 6]
```diff
@@ -483,18 +483,18 @@ public WebMonitorEndpoint(
...
+		final JobCancellationHandler jobCancelTerminationHandler = new JobCancellationHandler(
 			leaderRetriever,
 			timeout,
 			responseHeaders,
-			JobTerminationHeaders.getInstance(),
+			JobCancellationHeaders.getInstance(),
 			TerminationModeQueryParameter.TerminationMode.CANCEL);
 
-		final JobTerminationHandler jobStopTerminationHandler = new JobTerminationHandler(
+		final JobCancellationHandler jobStopTerminationHandler = new JobCancellationHandler(
```
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/WebMonitorEndpoint.java]: add a comment that this one is kept for legacy reasons

**Comment by kl0u** [comment]
I integrated your comments @zentol . Please have a look and let me know what you think.

File: flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/WebFrontendITCase.java [hunk 7]
```diff
@@ -280,7 +280,7 @@ public void testCancel() throws Exception {
 
 		try (HttpTestClient client = new HttpTestClient("localhost", getRestPort())) {
 			// cancel the job
-			client.sendPatchRequest("/jobs/" + jid + "/?mode=cancel", deadline.timeLeft());
+			client.sendPatchRequest("/jobs/" + jid + "/", deadline.timeLeft());
```
Comment (zentol) [inline flink-runtime-web/src/test/java/org/apache/flink/runtime/webmonitor/WebFrontendITCase.java]: remove trailing slash?

**Comment by kl0u** [comment]
Merged with e95b347dda5233f22fb03e408f2aa521ff924996
